File: core/handshape.py
"""Geometric handshape predicates over a single hand's 21 landmarks.

Pure 2D geometry, orientation-tolerant where possible. Each predicate returns a confidence in
[0, 1]. The verifier smooths these across recent frames (median) so one noisy frame can't flip a
result.

Shapes supported:
  fist / s    — four fingers curled, thumb unconstrained (S-hand / plain fist)
  a           — four fingers curled AND thumb extended alongside (letter A)
  index / 1   — index extended, other three curled (pointing / "1" hand)
  open / b / 5 — all four fingers extended (flat palm / B-hand)
  claw        — fingers clearly curled but not fully closed (E-hand / bent-5)
  point       — index extended, others curled (exact per-finger pattern; alias of index)
  v           — index + middle extended, ring + pinky curled (V / peace)
  l           — thumb + index extended, others curled (L)
  y           — thumb + pinky extended, others curled (Y)

fist/a/index/open/claw use averaged curl scoring (hospital scenario calibration); v/l/y/point use
an exact per-finger pattern match. Both are smoothed by the verifier across frames.
"""
from __future__ import annotations

import numpy as np

from core.landmarks import (
    Hand,
    WRIST,
    THUMB_TIP,
    INDEX_MCP,
    INDEX_TIP,
    MIDDLE_MCP,
    MIDDLE_TIP,
    RING_MCP,
    RING_TIP,
    PINKY_MCP,
    PINKY_TIP,
)

# (tip, mcp) per non-thumb finger
_FINGER_LM = {
    "index": (INDEX_TIP, INDEX_MCP),
    "middle": (MIDDLE_TIP, MIDDLE_MCP),
    "ring": (RING_TIP, RING_MCP),
    "pinky": (PINKY_TIP, PINKY_MCP),
}
_FINGERS = tuple(_FINGER_LM.values())


def _xy(hand: Hand, idx: int) -> np.ndarray:
    return hand.points[idx, :2]


def _hand_scale(hand: Hand) -> float:
    s = float(np.linalg.norm(_xy(hand, MIDDLE_MCP) - _xy(hand, WRIST)))
    return s if s > 1e-6 else 1.0


def _finger_curl(hand: Hand, tip: int, mcp: int) -> float:
    """1.0 = curled (tip folded toward palm), 0.0 = extended.

    Uses the ratio of (tip->wrist) to (mcp->wrist): an extended finger puts its tip far past the
    knuckle (ratio ~1.6+); a curled finger folds the tip back (ratio drops to ~1.0 or below).
    """
    tip_d = float(np.linalg.norm(_xy(hand, tip) - _xy(hand, WRIST)))
    mcp_d = float(np.linalg.norm(_xy(hand, mcp) - _xy(hand, WRIST)))
    r = tip_d / max(mcp_d, 1e-6)
    return float(np.clip((1.6 - r) / (1.6 - 1.0), 0.0, 1.0))


def _all_curls(hand: Hand) -> list[float]:
    return [_finger_curl(hand, t, m) for t, m in _FINGERS]


def _thumb_extended(hand: Hand) -> float:
    """1.0 = thumb sticking out alongside the hand, 0.0 = tucked/across the palm."""
    d = float(np.linalg.norm(_xy(hand, THUMB_TIP) - _xy(hand, INDEX_MCP))) / _hand_scale(hand)
    return float(np.clip((d - 0.5) / (1.2 - 0.5), 0.0, 1.0))


def extensions(hand: Hand) -> dict:
    """Per-digit extension in [0,1] (1 = extended, 0 = curled)."""
    ext = {name: 1.0 - _finger_curl(hand, tip, mcp) for name, (tip, mcp) in _FINGER_LM.items()}
    ext["thumb"] = _thumb_extended(hand)
    return ext


# --------------------------------------------------------------------------- averaged scorers
def fist_confidence(hand: Hand) -> float:
    """Four fingers curled (thumb unconstrained). Covers fist and S-handshape."""
    return float(np.mean(_all_curls(hand)))


def a_confidence(hand: Hand) -> float:
    """Letter A: four fingers curled AND thumb extended alongside (not wrapped across)."""
    return float(min(fist_confidence(hand), _thumb_extended(hand)))


def index_confidence(hand: Hand) -> float:
    """Index finger extended, the other three curled (1-hand / D / pointing)."""
    curls = _all_curls(hand)
    index_extended = 1.0 - curls[0]
    rest_curled = float(np.mean(curls[1:]))
    return float(np.clip(index_extended * 0.5 + rest_curled * 0.5, 0.0, 1.0))


def open_confidence(hand: Hand) -> float:
    """Open / flat hand: all four fingers extended (B-hand / flat palm / 5)."""
    return float(np.clip(1.0 - float(np.mean(_all_curls(hand))), 0.0, 1.0))


def claw_confidence(hand: Hand) -> float:
    """Fingers clearly curled but not fully closed (E-hand / bent-5 approximation).

    Used for MEDICINE and EMERGENCY. Generously scored; the repeated-motion detector carries the
    discriminating weight for those signs, so the handshape only confirms the hand is closed-ish.
    """
    curls = _all_curls(hand)
    m = float(np.mean(curls))
    base = float(np.clip((m - 0.25) / 0.35, 0.0, 1.0))   # 0 at flat, saturates ~0.60
    # A claw has ALL fingers similarly (partly) curled. A wide SPREAD of curls means some fingers
    # are fully out and some fully in — that's a finger-counting shape (n / w / index), not a claw.
    # Penalising spread stops a 2-finger "n" hand (mean curl ~0.5) from reading as a claw.
    spread = float(np.std(curls))
    penalty = float(np.clip(1.0 - max(0.0, spread - 0.15) / 0.35, 0.0, 1.0))
    return float(base * penalty)


def n_confidence(hand: Hand) -> float:
    """Two fingers (index + middle) extended, ring + pinky curled (the "N"/"U"/"H" family).

    v1 approximation: the real N/H thumb details aren't reliably detectable, so we recognise these
    by FINGER COUNT — exactly two extended fingers. NURSE and HOSPITAL use this; it is a minimal
    pair with DOCTOR (a flat hand), distinguished by how many fingers are extended.
    """
    c = _all_curls(hand)
    return float(np.clip(np.mean([1.0 - c[0], 1.0 - c[1], c[2], c[3]]), 0.0, 1.0))


def w_confidence(hand: Hand) -> float:
    """Three fingers (index + middle + ring) extended, pinky curled (the "W" / 3-hand). WATER."""
    c = _all_curls(hand)
    return float(np.clip(np.mean([1.0 - c[0], 1.0 - c[1], 1.0 - c[2], c[3]]), 0.0, 1.0))


def middle_confidence(hand: Hand) -> float:
    """Middle finger extended, the other three curled — the SICK "agony" handshape (v1 approx)."""
    c = _all_curls(hand)
    return float(np.clip(np.mean([c[0], 1.0 - c[1], c[2], c[3]]), 0.0, 1.0))


# --------------------------------------------------------------------------- exact patterns
# 1 = must be extended, 0 = must be curled, absent = don't care
_PATTERNS = {
    "point": dict(index=1, middle=0, ring=0, pinky=0),
    "1": dict(index=1, middle=0, ring=0, pinky=0),
    "v": dict(index=1, middle=1, ring=0, pinky=0),
    "l": dict(thumb=1, index=1, middle=0, ring=0, pinky=0),
    "y": dict(thumb=1, index=0, middle=0, ring=0, pinky=1),
}


def _match(hand: Hand, pattern: dict) -> float:
    ext = extensions(hand)
    scores = [ext[f] if target == 1 else 1.0 - ext[f] for f, target in pattern.items()]
    return float(min(scores)) if scores else 0.0


# --------------------------------------------------------------------------- dispatch
_DISPATCH = {
    "fist": fist_confidence,
    "s": fist_confidence,
    "a": a_confidence,
    "index": index_confidence,
    "open": open_confidence,
    "b": open_confidence,
    "5": open_confidence,
    "claw": claw_confidence,
    "n": n_confidence,              # 2 fingers — NURSE
    "h": n_confidence,              # H is the same 2-finger shape — HOSPITAL
    "u": n_confidence,              # alias
    "w": w_confidence,              # 3 fingers — WATER
    "middle": middle_confidence,    # SICK
}


def handshape_confidence(hand: Hand, kind: str) -> float:
    """Confidence in [0, 1] that `hand` forms handshape `kind`. Unknown kinds score 0."""
    kind = kind.lower()
    fn = _DISPATCH.get(kind)
    if fn is not None:
        return fn(hand)
    pattern = _PATTERNS.get(kind)
    return _match(hand, pattern) if pattern is not None else 0.0

File: web/src/engine/handshape.ts
import { clip, mean, std } from './math-utils';
import type { Hand } from './landmarks';
import { WRIST, THUMB_TIP, INDEX_MCP, INDEX_TIP, MIDDLE_MCP, MIDDLE_TIP, RING_MCP, RING_TIP, PINKY_MCP, PINKY_TIP } from './landmarks';

type FingerPair = [number, number]; // [tip, mcp]
const FINGER_LM: Record<string, FingerPair> = {
  index: [INDEX_TIP, INDEX_MCP],
  middle: [MIDDLE_TIP, MIDDLE_MCP],
  ring: [RING_TIP, RING_MCP],
  pinky: [PINKY_TIP, PINKY_MCP],
};
const FINGERS: FingerPair[] = Object.values(FINGER_LM);

function xy(hand: Hand, idx: number): [number, number] {
  return [hand.points[idx][0], hand.points[idx][1]];
}

function dist2d(a: [number, number], b: [number, number]): number {
  const dx = a[0] - b[0];
  const dy = a[1] - b[1];
  return Math.sqrt(dx * dx + dy * dy);
}

function handScale(hand: Hand): number {
  const s = dist2d(xy(hand, MIDDLE_MCP), xy(hand, WRIST));
  return s > 1e-6 ? s : 1.0;
}

function fingerCurl(hand: Hand, tip: number, mcp: number): number {
  const tipD = dist2d(xy(hand, tip), xy(hand, WRIST));
  const mcpD = dist2d(xy(hand, mcp), xy(hand, WRIST));
  const r = tipD / Math.max(mcpD, 1e-6);
  return clip((1.6 - r) / (1.6 - 1.0), 0, 1);
}

function allCurls(hand: Hand): number[] {
  return FINGERS.map(([t, m]) => fingerCurl(hand, t, m));
}

function thumbExtended(hand: Hand): number {
  const d = dist2d(xy(hand, THUMB_TIP), xy(hand, INDEX_MCP)) / handScale(hand);
  return clip((d - 0.5) / (1.2 - 0.5), 0, 1);
}

export function extensions(hand: Hand): Record<string, number> {
  const ext: Record<string, number> = {};
  for (const [name, [tip, mcp]] of Object.entries(FINGER_LM)) {
    ext[name] = 1.0 - fingerCurl(hand, tip, mcp);
  }
  ext.thumb = thumbExtended(hand);
  return ext;
}

function fistConfidence(hand: Hand): number {
  return mean(allCurls(hand));
}

function aConfidence(hand: Hand): number {
  return Math.min(fistConfidence(hand), thumbExtended(hand));
}

function indexConfidence(hand: Hand): number {
  const curls = allCurls(hand);
  const indexExtended = 1.0 - curls[0];
  const restCurled = mean(curls.slice(1));
  return clip(indexExtended * 0.5 + restCurled * 0.5, 0, 1);
}

function openConfidence(hand: Hand): number {
  return clip(1.0 - mean(allCurls(hand)), 0, 1);
}

function clawConfidence(hand: Hand): number {
  const curls = allCurls(hand);
  const m = mean(curls);
  const base = clip((m - 0.25) / 0.35, 0, 1);
  const spread = std(curls);
  const penalty = clip(1.0 - Math.max(0, spread - 0.15) / 0.35, 0, 1);
  return base * penalty;
}

function nConfidence(hand: Hand): number {
  const c = allCurls(hand);
  return clip(mean([1.0 - c[0], 1.0 - c[1], c[2], c[3]]), 0, 1);
}

function wConfidence(hand: Hand): number {
  const c = allCurls(hand);
  return clip(mean([1.0 - c[0], 1.0 - c[1], 1.0 - c[2], c[3]]), 0, 1);
}

function middleConfidence(hand: Hand): number {
  const c = allCurls(hand);
  return clip(mean([c[0], 1.0 - c[1], c[2], c[3]]), 0, 1);
}

const PATTERNS: Record<string, Record<string, number>> = {
  point: { index: 1, middle: 0, ring: 0, pinky: 0 },
  '1': { index: 1, middle: 0, ring: 0, pinky: 0 },
  v: { index: 1, middle: 1, ring: 0, pinky: 0 },
  l: { thumb: 1, index: 1, middle: 0, ring: 0, pinky: 0 },
  y: { thumb: 1, index: 0, middle: 0, ring: 0, pinky: 1 },
  // strict min-based: averaged version let open hands score 0.5+
  n: { index: 1, middle: 1, ring: 0, pinky: 0 },
  h: { index: 1, middle: 1, ring: 0, pinky: 0 },
  u: { index: 1, middle: 1, ring: 0, pinky: 0 },
  w: { index: 1, middle: 1, ring: 1, pinky: 0 },
  middle: { index: 0, middle: 1, ring: 0, pinky: 0 },
};

function matchPattern(hand: Hand, pattern: Record<string, number>): number {
  const ext = extensions(hand);
  const scores = Object.entries(pattern).map(([f, target]) =>
    target === 1 ? ext[f] : 1.0 - ext[f]
  );
  return scores.length > 0 ? Math.min(...scores) : 0;
}

const DISPATCH: Record<string, (hand: Hand) => number> = {
  fist: fistConfidence,
  s: fistConfidence,
  a: aConfidence,
  index: indexConfidence,
  open: openConfidence,
  b: openConfidence,
  '5': openConfidence,
  claw: clawConfidence,
  n: nConfidence,
  w: wConfidence,
  middle: middleConfidence,
};

export function handshapeConfidence(hand: Hand, kind: string): number {
  const k = kind.toLowerCase();
  const fn = DISPATCH[k];
  if (fn) return fn(hand);
  const pattern = PATTERNS[k];
  if (pattern) return matchPattern(hand, pattern);
  return 0;
}
